### PR TITLE
ramips: add missing information to dlink headers

### PR DESCRIPTION
**Pull request by Lucky1openwrt**
This add header information needed so some of
the newer bootloader found in DIR-2660-A1 & A2
and a step closer to a web flash image
also fix a mislabelled flash partition type
by removing 2nd kernel & UBI partition

Signed-off-by: Alan Luck <luckyhome2008@gmail.com>

### CODE REVIEW

**Comment by adschm** [comment]
@bmork FYI

**Comment by basrieter** [comment]
I experience the same issues and can confirm these changes resolve the flashing problems.

**Comment by bmork** [comment]
Sorry about that.  Should have fixed it uo while messing with the partition splitter.  But I wonder:  Did the behaviour change with my mtdsplit changes?  I believe it should not.

Anyway, we can't currently have two partitions with automatic rootfs splitting AFAIK. To make that work, the splitter would have to be aware of which firmware partition the kernel was loaded from.

One final, and somewhat unrelated, note:  I believe the "read-only" property on secondary firmware partitions, or other non-critical partitions, does more harm than good.  It's not like we do arbitrary writes to the partition if it is writable.  And worst case if we do is still a bootable system, since OpenWrt boots from the primary partition anyway. The problem with having read-only is that it makes it really hard to fix anything if the bootloader decides to boot the "wrong" partition for some reason.  We have plenty of experience with this from the Unifi AP AC series.  In theory, you can just install the kmod-mtd-rw package to override the read-only,  But if the problem is a kernel vs rootfs mismatch, which often will be the issue, then that isn't straight forward.

So IMHO, read-only should be reserved for those few partitions we never want to modify and which are critical for the system, like the bootloader, device specific factory configuration, and calibration data.

**Comment by Lucky1openwrt** [comment]
I didn't notice any change with mtdsplit
but I did think that while sysupgrade was getting confused when keeping setting 
but didn't pin it down to anything 
but think that having 2 kernel & UBI's was confusing maybe me more then It

this unit only boot's from firmware but the boot loader uses firmware2 as a backup
if it's uboot header checksum is valid & different from firmware2
it only backs up the amount of bytes in the uboot header but for openwrt it's not the whole thing
but if the main firmware checksum fails it copy's it back from firmware2

about read only 
well they can be removed but kernal2 the boot loader has control over the start + size of 1st partition in uboot header

due to the DTSI covering a few models it's had to know which are used on witch models
but "private" that follows firmware on at lest the DIR-1960 is all "FF"s & unused 
on my unit I have added it to firmware to make the room bigger but don't know about the others models
there as another unused reserved at the end also

I did try to setup concat partitioning but failed to get it to work 
always failed on 2nd boot not being able to find the UBI partation so I gave up

but these 2 partition@2980000 & partition@7780000 could also be used for other things

**Comment by bmork** [comment]
> this unit only boot's from firmware but the boot loader uses firmware2 as a backup
> if it's uboot header checksum is valid & different from firmware2
> it only backs up the amount of bytes in the uboot header but for openwrt it's not the whole thing
> but if the main firmware checksum fails it copy's it back from firmware2

That sounds familiar.  I am currently trying to add support for the ZyXEL NR7101, which also is based on MT7621.  It has exactly the same behaviour.  I believe it comes from the chipset SDK and therefore is a cross-vendor thing.

I tried to describe my observations of the scheme in the draft commit message.  Scroll down to the section titled "Notes on the OEM/bootloader dual partition scheme" on
https://patchwork.ozlabs.org/project/openwrt/patch/20210419110056.6348-3-bjorn@mork.no/

The short version is that it's hard to make any good use of this scheme in OpenWrt. The best I've come up with is installing a working initramfs in the second partition.  That will at least work as an emergency recovery installation.

We could reserve only a smaller part of the secondary partition for that recovery initramfs, with a fixed split, and use the rest for an additional UBI.  But I'm not sure the additional complexity is worth it.  And we'd have to make sure the reserved part is large enough to hold any kernel image we install in the primary partition, in case the bootloader decides to copy it to the secondary.

**Comment by Lucky1openwrt** [comment]
yes was thinking of fixing the kernel size so if it got bigger someone would have to change it manually
& so they would change the same in the secondary partition
but after failing to get concat working i just let it as is

**Comment by kar200** [comment]
I confirm this works  for my device as well (DIR-853-A3) which is not part of openwrt yet.

Since the device does not have a way to boot into the emergency mode with the reset button (or any other button I tried except through console). I just have to compile openwrt with the patch included - encrypt the image with imgcrypt (a 32 bit binary found in the dlink gpl source code) and just flash it directly from the stock Dlink web interface.

File: target/linux/ramips/image/mt7621.mk [hunk 1]
```diff
@@ -237,12 +237,11 @@ define Device/dlink_dir-8xx-a1
...
-  KERNEL_INITRAMFS := $$(KERNEL) | uimage-padhdr 96
+  KERNEL := $$(KERNEL) | uimage-sgehdr
   IMAGES += factory.bin
-  IMAGE/sysupgrade.bin := append-kernel | append-rootfs | uimage-padhdr 96 |\
-	pad-rootfs | append-metadata | check-size
-  IMAGE/factory.bin := append-kernel | append-rootfs | uimage-padhdr 96 |\
-	check-size
+  IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs |\
+  	append-metadata | check-size
+  IMAGE/factory.bin := append-kernel | append-rootfs |	check-size
```
Comment (adschm) [inline target/linux/ramips/image/mt7621.mk:244]: Whitespace issue here.

File: target/linux/ramips/image/mt7621.mk [hunk 2]
```diff
@@ -237,12 +237,11 @@ define Device/dlink_dir-8xx-a1
...
   DEVICE_PACKAGES := kmod-mt7615e kmod-mt7615-firmware
-  KERNEL_INITRAMFS := $$(KERNEL) | uimage-padhdr 96
+  KERNEL := $$(KERNEL) | uimage-sgehdr
   IMAGES += factory.bin
-  IMAGE/sysupgrade.bin := append-kernel | append-rootfs | uimage-padhdr 96 |\
-	pad-rootfs | append-metadata | check-size
-  IMAGE/factory.bin := append-kernel | append-rootfs | uimage-padhdr 96 |\
-	check-size
+  IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs |\
+  	append-metadata | check-size
```
Comment (adschm) [inline target/linux/ramips/image/mt7621.mk:243]: Hanging indent should be one tab only.

File: tools/firmware-utils/src/uimage_sgehdr.c [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+ * Copyright (C) 2019 NOGUCHI Hiroshi <drvlabo@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License,
+ * version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
```
Comment (adschm) [inline tools/firmware-utils/src/uimage_sgehdr.c:14]: Please use SPDX license identifier instead.

File: target/linux/ramips/image/mt7621.mk (hunk 1 above)
Comment (Lucky1openwrt) [reply target/linux/ramips/image/mt7621.mk]: I think it's fixed

File: target/linux/ramips/image/mt7621.mk (hunk 2 above)
Comment (Lucky1openwrt) [reply target/linux/ramips/image/mt7621.mk]: not sure may have fixed it

File: target/linux/ramips/image/mt7621.mk [hunk 4]
```diff
@@ -273,10 +272,10 @@ define Device/dlink_dir-xx60-a1
   DEVICE_VENDOR := D-Link
   DEVICE_PACKAGES := kmod-mt7615e kmod-mt7615-firmware kmod-usb3 \
 	kmod-usb-ledtrig-usbport
-  KERNEL := $$(KERNEL) | uimage-padhdr 96
+  KERNEL := $$(KERNEL) | uimage-sgehdr
   IMAGES += factory.bin
   IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
-  IMAGE/factory.bin := append-kernel | pad-to $$(KERNEL_SIZE) | append-ubi | \
+  IMAGE/factory.bin := append-kernel | pad-to $$(KERNEL_SIZE) | append-ubi |\
```
Comment (adschm) [inline target/linux/ramips/image/mt7621.mk:278]: This change seems unneeded and unrelated.

Comment (Lucky1openwrt) [reply target/linux/ramips/image/mt7621.mk:278]: was the only "Whitespace" thing i could find

**Comment by basrieter** [comment]
Any update on this? The device is actually really powerful and I would really like to start using it. But it would be great to have proper official images for it.

**Comment by Lucky1openwrt** [comment]
I think it's just going to sit here

**Comment by Lucky1openwrt** [comment]
here is an rc2 with the header manually changed
http://luckys.onmypc.net/openwrt/DIR-2660/openwrt-21.02.0-rc2-ramips-mt7621-dlink_dir-2660-a1-squashfs-factory.bin

**Comment by basrieter** [comment]
Thanks @Lucky1openwrt! 

@adschm What is needed to get this moving again?

**Comment by adschm** [comment]
Well, first the remaining comment needs to be addressed.

Apart from that, people that review the proposed change.

File: target/linux/ramips/image/mt7621.mk (hunk 4 above)
Comment (basrieter) [reply target/linux/ramips/image/mt7621.mk:278]: So this needs to become `IMAGE/factory.bin := append-kernel | pad-to $$(KERNEL_SIZE) | append-ubi | \`

Comment (basrieter) [reply target/linux/ramips/image/mt7621.mk:278]: So with a space between `|` and `\`.

Comment (adschm) [reply target/linux/ramips/image/mt7621.mk:278]: Yes, since this would mean it's _not changed_.

**Comment by CodeFetch** [comment]
@s-2 Can you please have a look?

**Comment by s-2** [comment]
> @s-2 Can you please have a look?

saw this one already (see my thumbs up in first post), can't say much about the uimage header though (this is not directly related to factory image encryption, but the payload itself), but I'm happy to see someone take care of these devices =)

**Comment by basrieter** [comment]
So if the whitespace issue is fixed, this could go in?

File: target/linux/ramips/image/mt7621.mk (hunk 4 above)
Comment (Lucky1openwrt) [reply target/linux/ramips/image/mt7621.mk]: add space

**Comment by basrieter** [comment]
Anyone willing to look into this and approve? @Lucky1openwrt  fixed the whitespace issue.

File: target/linux/ramips/image/Makefile [hunk 5]
```diff
@@ -144,6 +144,11 @@ define Build/uimage-padhdr
 	mv $@.new $@
 endef
 
+define Build/uimage-sgehdr
+	uimage_sgehdr -i $@ -o $@.new  -m $(DEVICE_MODEL) -h $(DEVICE_VARIANT) -s V1.00000
```
Comment (adschm) [inline target/linux/ramips/image/Makefile:148]: Nitpick: two spaces after $@.new ...

Comment (adschm) [reply target/linux/ramips/image/Makefile:148]: Probably introduced when removing the line wrap that should be here.

Comment (Lucky1openwrt) [reply target/linux/ramips/image/Makefile]: removed space & added line wrap

Comment (basrieter) [reply target/linux/ramips/image/Makefile]: @adschm I guess the whitespace/line wrap issue is fixed now?

**Comment by basrieter** [comment]
Is this mergeable now? I really would like to have sysupgrades for this nice dual cpu router?

**Comment by Lucky1openwrt** [comment]
I have done all I can

**Comment by basrieter** [comment]
@adschm Is there any way we can get this moved? @Lucky1openwrt made the changes with the whitespaces?

**Comment by kar200** [comment]
Hi @Lucky1openwrt 

Can you please add the DIR-853-A3 in the code?

**Comment by Lucky1openwrt** [comment]
sure will do soon :)

**Comment by Lucky1openwrt** [comment]
nothing specific need for the dir-853-a3
all covered under "dlink_dir-xx60-a1" image creation

**Comment by Lucky1openwrt** [comment]
not sure this will make it's way through to 21.02.1 now

**Comment by kar200** [comment]
> nothing specific need for the dir-853-a3
> all covered under "dlink_dir-xx60-a1" image creation

Ah yeah it is indeed. By the way I tested this with the encryption from the other pull request and the image generated flashed from the web interface with no issues.

**Comment by basrieter** [comment]
> not sure this will make it's way through to 21.02.1 now

That would really be a shame. Isn't there anyone that can merge this? Already a couple of users that are reporting success now.

**Comment by gnoymmij** [comment]
Yes, it would really be a shame if this doesn't get merged. I am really hoping the next RC or final release would officially support this amazing router.

**Comment by s-2** [comment]
usually device support is not backported to the current release branch, so we can expect this to be in publicly available images by the end of 2022 (let's hope the devices will still be in production by then), however this one might be considered a minor fix for existing devices to ensure flashability at all, and thus should see much higher chances of backporting :+1:

**Comment by basrieter** [comment]
Waiting till 2022 would be a real shame. Ad this router is currently on of the more powerfull routers.

So the question remains: who can make this merge?

**Comment by s-2** [comment]
> who can make this merge?

Technically, any of the [developers](https://openwrt.org/about#people), however not everyone is familiar with all of the targets etc., and consider that no one is being paid for taking the time for reviewing and merging PRs; in the end everyone is just doing this in their personal free time.

**Comment by gnoymmij** [comment]
Based on the discussions here, seems like @adschm is our only hope of making this merge.

**Comment by basrieter** [comment]
And hopefully we can then do a PR for backporting? Or should we already create that @s-2 ? 

BTW: I really do appreciate all the hard work here. Being a OS dev myself, I know the (time) challenges.

**Comment by s-2** [comment]
> And hopefully we can then do a PR for backporting? Or should we already create that @s-2 ?

Maybe it can be prepared locally, or even in public branch, but I would wait for any further review comments first, before opening another PR.

**Comment by Lucky1openwrt** [comment]
?

**Comment by blocktrron** [comment]
Checkpatch still complained about various style problems.

```
WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
#110: 
new file mode 100644

WARNING: please, no space before tabs
#136: FILE: tools/firmware-utils/src/uimage_sgehdr.c:22:
+#define SGE_PRODUCTLEN ^I64^I/* sge_Product Length^I^I*/$

WARNING: please, no space before tabs
#137: FILE: tools/firmware-utils/src/uimage_sgehdr.c:23:
+#define SGE_VERSIONLEN ^I16^I/* sge Version Length^I^I*/$

WARNING: 'Orignal' may be misspelled - perhaps 'Original'?
#138: FILE: tools/firmware-utils/src/uimage_sgehdr.c:24:
+#define OrignalHL		64	/* Orignal Header Length	*/
                  		  	   ^^^^^^^

WARNING: do not add new typedefs
#144: FILE: tools/firmware-utils/src/uimage_sgehdr.c:30:
+typedef struct image_header {

WARNING: unnecessary whitespace before a quoted newline
#170: FILE: tools/firmware-utils/src/uimage_sgehdr.c:56:
+		"%s -i <input_uimage_file> -o <output_file> -m <model> -h <hardware version> -s <software version> \n",

ERROR: trailing whitespace
#265: FILE: tools/firmware-utils/src/uimage_sgehdr.c:151:
+^I$

ERROR: trailing whitespace
#269: FILE: tools/firmware-utils/src/uimage_sgehdr.c:155:
+^I$

total: 2 errors, 6 warnings, 230 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

NOTE: Whitespace errors detected.
      You may wish to use scripts/cleanpatch or scripts/cleanfile

0001-ramips-add-missing-information-to-dlink-headers.patch has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.
```

I've fixed up these errors to speed up the process.

@bmork I've removed the read-only flag removals from the commit message. I understand your concerns, however I'd like to hear more opinions on this matter. Before this is becoming a "guideline" (and the next person claims you can extend the rootfs by using a adjacent partition).

@Lucky1openwrt can you compile a image from my staging tree and confirm it works as intended on your device (If someone else wants to do this, this is also fine)

https://git.openwrt.org/?p=openwrt/staging/blocktrron.git;a=summary

Best

**Comment by Lucky1openwrt** [comment]
visual inspection & comparison with dlink firmware show there & in the correct place
after encrypting output file manually with "dlink-sge-image" 
from https://github.com/openwrt/openwrt/pull/4174
on a DIR-878 allows direct flashing from dlink firmware v1.3 via it's web interface

**Comment by basrieter** [comment]
As 21.02.0 has been release, I guess the chance for this device to be include will be almost zero?

**Comment by Neustradamus** [comment]
It can be added in 21.02.1 and master too, no?

**Comment by Lucky1openwrt** [comment]
it is a fix but who knows

**Comment by basrieter** [comment]
> it is a fix but who knows

I really hope so, as this is a really nice dual core router with a lot of potential.

**Comment by Lucky1openwrt** [comment]
thanks @blocktrron

**Comment by Neustradamus** [comment]
Merged commit is:
- https://github.com/openwrt/openwrt/commit/2c9a07ed28e87bbe2d221d48aaa19b04672646a5

**Comment by basrieter** [comment]
Nice. So we can now flash the development builds? And use sysupgrades?

**Comment by Lucky1openwrt** [comment]
just in the master branch tho

**Comment by basrieter** [comment]
> just in the master branch tho

Super. So will it appear as D-Link DIR-2660 A1 or D-Link DIR-2660 A2.

**Comment by Lucky1openwrt** [comment]
I think they where all tagged with A1
it's the only one there
https://downloads.openwrt.org/snapshots/targets/ramips/mt7621/
https://downloads.openwrt.org/snapshots/targets/ramips/mt7621/openwrt-ramips-mt7621-dlink_dir-2660-a1-squashfs-factory.bin

**Comment by gnoymmij** [comment]
Does this work with the original A1 hardware as well?

**Comment by Lucky1openwrt** [comment]
if your uboot needs it then yes
early models didn't but later one's do

**Comment by gnoymmij** [comment]
Meaning it is not backward compatible with early A1 hardware?

**Comment by Lucky1openwrt** [comment]
meaning it's was not needed on the earlier boot loaders
but will work fine it just fills in fields to match the dlink firmware

**Comment by ads119** [comment]
Hi,

I spent 20 hours trying to find more information on how to make a D-link DIR 853 A2 device work with OpenWRT and I still couldn't, sorry if I apologize if I'm not searching right, but can you help me find it, I need make this equipment urgently.

Thanks in advance for all your help.

**Comment by Lucky1openwrt** [comment]
you need to capture the boot log from the serial console 
to get information on flash chips partition layouts  and hardware
and it not going to be quick '
I don't think the radios that are in the a3 & a1 are work very well anyway
the switch in these devices is only now getting lots of work done to it
even these devices that are listed to have supper are well not that good atm

can't even find inf on the A2 and they differ in different parts of the world

**Comment by basrieter** [comment]
@Lucky1openwrt  I just tried the latest 21.02.2 build and notice that the `factory` image does have the new headers for the A2, but the `sysupgrade` does not. https://firmware-selector.openwrt.org/?version=21.02.2&target=ramips%2Fmt7621&id=dlink_dir-2660-a1.

The weird thing is that the snapshot versions do contain the headers in `sysupgrade` images.

**Comment by Lucky1openwrt** [comment]
the fix is only in the master branch
I don't know if the sysupgrade image needs it or not as i don't have this model and can't test it
I know it is needed for the upload of the factory image from the recovery page

**Comment by Lucky1openwrt** [comment]
you are welcome to make a pull request to v21 but I imaging v22 well be out before it's even looked at

**Comment by basrieter** [comment]
What is the risk if I try and the headers do not match?

**Comment by Lucky1openwrt** [comment]
if it's not checked on boot then sydupgrade image will just work
if it checks & failed then you have to upload a factory image again via recovery
it may even just go to the recover page

**Comment by basrieter** [comment]
> if it's not checked on boot then sydupgrade image will just work if it checks & failed then you have to upload a factory image again via recovery it may even just go to the recover page

I think I tested it way back and it will go to recovery. So I should wait for 22 or use a snapshot (which gave me issues with `opkg update` in the past.

**Comment by Lucky1openwrt** [comment]
i just tested the current master to test the new switch driver
but the wifi driver is bad & only seems to work at 20Mhz download
super slow that was an EA8100 linksys

**Comment by basrieter** [comment]
The master = snapshot right? So I should really stick to 21.02.2 then?

**Comment by Lucky1openwrt** [comment]
in this case the master only has snapshots 
but there are 21.02 snapshot's and v19's as well

**Comment by Lucky1openwrt** [comment]
I'll manually fix the header in V2102.2 when i get a chance

**Comment by basrieter** [comment]
I really wish they would back port it to 21.02.x

**Comment by basrieter** [comment]
> i just tested the current master to test the new switch driver but the wifi driver is bad & only seems to work at 20Mhz download super slow that was an EA8100 linksys still MT7621 & MT7615N's

So the WIFI driver sucks in the 22.x.x snapshots.

**Comment by ads119** [comment]
> você precisa capturar o log de inicialização do console serial para obter informações sobre layouts de partição de chips flash e hardware e não será rápido ' não acho que os rádios que estão no a3 e a1 funcionem muito bem de qualquer maneira o switch nestes dispositivos só agora está sendo feito muito trabalho, mesmo esses dispositivos listados para jantar não são tão bons assim
> 
> nem consigo encontrar informações no A2 e eles diferem em diferentes partes do mundo

**Comment by ads119** [comment]
This my device:
Model Name: DIR-853 Hardware Version: A2
Firmware Version: 1.20

I accessed the router via Wireless (http://192.168.0.1) and managed to get this log below... Does it help to understand how I should proceed to make Open WRT work on this model of equipment? Sorry, I don't quite understand yet.


2020-07-14 12:55:40 [SYSLOG]: start BusyBox v1.12.1
2020-07-14 12:55:40 [MLD]: mrib_mld_source: failed to detect local source address on 13
2020-07-14 12:55:40 [MLD]: mrib_mld_source: failed to detect local source address on 15
2020-07-14 12:55:40 [MLD]: mrib_mld_source: failed to detect local source address on 16
2020-07-14 12:56:07 [RC]: RC start
              total         used         free       shared      buffers
  Mem:       120980        79256        41724            0            0
 Swap:            0            0            0
Total:       120980        79256        41724
  PID USER       VSZ STAT COMMAND
    1 admin     4784 R    /sbin/preinit 
    2 admin        0 SW   [kthreadd]
    3 admin        0 SW   [ksoftirqd/0]
    4 admin        0 SW   [kworker/0:0]
    5 admin        0 SW<  [kworker/0:0H]
    6 admin        0 SW   [kworker/u8:0]
    7 admin        0 SW   [migration/0]
    8 admin        0 SW   [rcu_bh]
    9 admin        0 SW   [rcu_sched]
   10 admin        0 SW   [migration/1]
   11 admin        0 SW   [ksoftirqd/1]
   12 admin        0 SW   [kworker/1:0]
   13 admin        0 SW<  [kworker/1:0H]
   14 admin        0 SW   [migration/2]
   15 admin        0 SW   [ksoftirqd/2]
   16 admin        0 SW   [kworker/2:0]
   17 admin        0 SW<  [kworker/2:0H]
   18 admin        0 SW   [migration/3]
   19 admin        0 SW   [ksoftirqd/3]
   20 admin        0 SW   [kworker/3:0]
   21 admin        0 SW<  [kworker/3:0H]
   22 admin        0 SW<  [khelper]
   23 admin        0 SW   [kdevtmpfs]
   24 admin        0 SW<  [netns]
   25 admin        0 SW<  [writeback]
   26 admin        0 SW<  [bioset]
   27 admin        0 SW<  [kblockd]
   28 admin        0 SW   [khubd]
   29 admin        0 SW   [kworker/3:1]
   30 admin        0 SW   [kworker/2:1]
   31 admin        0 SW   [kworker/1:1]
   32 admin        0 SW   [kworker/0:1]
   33 admin        0 SW   [kswapd0]
   34 admin        0 SWN  [ksmd]
   35 admin        0 SW   [fsnotify_mark]
   36 admin        0 SW<  [crypto]
   43 admin        0 SW   [kworker/u8:1]
   48 admin        0 SW<  [deferwq]
   49 admin        0 SWN  [jffs2_gcd_mtd6]
   57 admin     1132 S    tw_hotplug 
   63 admin     5592 S    nvram_daemon 
  439 admin     3972 S    stad 1 
  441 admin     3976 S    stad 2 
  454 admin     4784 S    /sbin/preinit 
  542 admin      944 S    nl_server -i br0 -s dlinkrouter -s dlinkrouterEEBA -s
  543 admin      932 S    nl_server -i br0 -s dlinkrouter -s dlinkrouterEEBA -s
  549 admin     1424 S    dnsmasq -C /tmp/dnsmasq.conf -6 /tmp/dnsmasq_dhcp_eve
  550 admin     1416 S    dnsmasq -C /tmp/dnsmasq.conf -6 /tmp/dnsmasq_dhcp_eve
  570 admin     4804 S    /bin/lighttpd -f /etc_ro/lighttpd/lighttpd.conf -m /e
  571 admin     5572 S    /bin/prog.cgi 
  574 admin     1904 S    dhcp6s -c /tmp/dhcp6s.conf -k /tmp/dhcp6sctlkey br0 
  576 admin     1032 S    radvd -m none -C /tmp/radvd.conf 
  584 admin     1700 S    syslogd -L 
  591 admin      980 S    omcproxy -e3 eth3 br0 br2 br3 scope=realm 
  639 admin        0 SW   [RtmpCmdQTask]
  640 admin        0 SW   [RtmpWscTask]
  641 admin        0 SW   [HwCtrlTask]
  642 admin        0 SW   [ser_task]
  643 admin        0 SW   [kworker/1:2]
  660 admin        0 SW   [RtmpMlmeTask]
  709 admin     1464 S    bndstrg2 -i ra0 -i rax0 
 1164 admin     1320 S    mDNSResponder -b -i br0 -f /tmp/mdns_resp.conf -e dli
 1219 admin     4144 S    timer 
 1847 admin     1704 S    sh -c ps >> /etc_ro/lighttpd/www/web/messages 
 1848 admin     1704 R    ps 
br0       Link encap:Ethernet  HWaddr C4:E9:0A:4C:EE:BA  
          inet addr:192.168.0.1  Bcast:192.168.0.255  Mask:255.255.255.0
          inet6 addr: fe80::c6e9:aff:fe4c:eeba/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:1964 errors:0 dropped:0 overruns:0 frame:0
          TX packets:681 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:143648 (140.2 KiB)  TX bytes:654412 (639.0 KiB)

br1       Link encap:Ethernet  HWaddr C4:E9:0A:4C:EE:BE  
          inet addr:192.168.7.1  Bcast:192.168.7.255  Mask:255.255.255.0
          inet6 addr: fe80::603e:50ff:fe8c:37e5/64 Scope:Link
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:4 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:0 (0.0 B)  TX bytes:408 (408.0 B)

br2       Link encap:Ethernet  HWaddr C4:E9:0A:4C:EE:BF  
          inet6 addr: fe80::f4de:acff:fed9:db6c/64 Scope:Link
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:6 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:0 (0.0 B)  TX bytes:568 (568.0 B)

br3       Link encap:Ethernet  HWaddr C4:E9:0A:4C:EE:C0  
          inet6 addr: fe80::f0b2:a7ff:fe31:b11/64 Scope:Link
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:6 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:0 (0.0 B)  TX bytes:568 (568.0 B)

eth2      Link encap:Ethernet  HWaddr C4:E9:0A:4C:EE:BA  
          inet6 addr: fe80::c6e9:aff:fe4c:eeba/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:1741 errors:0 dropped:0 overruns:0 frame:0
          TX packets:2805 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:156121 (152.4 KiB)  TX bytes:309541 (302.2 KiB)
          Interrupt:3 

eth2.1    Link encap:Ethernet  HWaddr C4:E9:0A:4C:EE:BA  
          inet6 addr: fe80::c6e9:aff:fe4c:eeba/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:265 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:0 (0.0 B)  TX bytes:37189 (36.3 KiB)

eth2.2    Link encap:Ethernet  HWaddr C4:E9:0A:4C:EE:BA  
          inet6 addr: fe80::c6e9:aff:fe4c:eeba/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:266 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:0 (0.0 B)  TX bytes:37259 (36.3 KiB)

eth2.3    Link encap:Ethernet  HWaddr C4:E9:0A:4C:EE:BA  
          inet6 addr: fe80::c6e9:aff:fe4c:eeba/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:265 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:0 (0.0 B)  TX bytes:37189 (36.3 KiB)

eth2.4    Link encap:Ethernet  HWaddr C4:E9:0A:4C:EE:BA  
          inet6 addr: fe80::c6e9:aff:fe4c:eeba/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:265 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:0 (0.0 B)  TX bytes:37189 (36.3 KiB)

eth3      Link encap:Ethernet  HWaddr C4:E9:0A:4C:EE:BD  
          inet6 addr: fe80::c6e9:aff:fe4c:eebd/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:3 errors:0 dropped:0 overruns:0 frame:0
          TX packets:5 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:238 (238.0 B)  TX bytes:558 (558.0 B)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

ra0       Link encap:Ethernet  HWaddr C4:E9:0A:4C:EE:BC  
          inet6 addr: fe80::c6e9:aff:fe4c:eebc/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
          Interrupt:4 

rax0      Link encap:Ethernet  HWaddr C4:E9:0A:4C:EE:BB  
          inet6 addr: fe80::c6e9:aff:fe4c:eebb/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:2089 errors:0 dropped:0 overruns:0 frame:0
          TX packets:938 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:177562 (173.4 KiB)  TX bytes:652296 (637.0 KiB)

**Comment by Lucky1openwrt** [comment]
> So the WIFI driver sucks in the 22.x.x snapshots.
the wifi driver in the Master snapshot didn't work 2 day's ago for me well stuck at 20Mhz only
they may have fixed it by now it's a day later
that's all i can tell you

they have not branched v22 yet but i think it's soon

just to clarify a Snapshot is a compilation of the current stated of each branch
there is a Master Snapshot and a V21.02 Snapshot & I think even still a V19.07 snapshot

here is the 21.02.2 factory with the header modification 
http://luckys.onmypc.net/openwrt/DIR-2660/openwrt-21.02.2-ramips-mt7621-dlink_dir-2660-a1-squashfs-factory.bin

**Comment by basrieter** [comment]
Thanks! Is it possible to also create a sysupgrade image?

**Comment by Lucky1openwrt** [comment]
> Thanks! Is it possible to also create a sysupgrade image?
I'm sure it is but i don't know atm
you could try just replacing the file in the image with winzip
